### PR TITLE
Rewrite as Web extension

### PR DESCRIPTION
**Pull request by est31**
Proof of concept for a WebExtension rewrite.

cc #23

### CODE REVIEW

**Comment by est31** [comment]
So currently it looks like:

1. it guesses whether a page will yield a prompt and if yes, it redirects to a page controlled by us
2. that page then shows a question whether to a) download the file b) open it in browser or c) do the default action
3. the page lets the background script execute the user action

It doesn't work for the download method by setting the MIME types to application/octet-stream yet, its not translated yet, and its terribly ugly. But as proof of concept it works.

Also, adding an entry to the native dialog would be far much better. But whatever, this is the only option we have got (until the firefox team implements it).

**Comment by est31** [comment]
Also, I've implemented no custom MIME type setting yet.

**Comment by spasche** [comment]
Thank you for having a look at this.

I don't have much time / motivation to work on this extension these days. Because of that and the fact that the new WebExtension version might be more or less a full rewrite, it could be more efficient to let it live in its own repository if you're interested to continue working on it.

Once you have a working WebExtension version I'd be happy to give you write access to addons.mozilla.org to publish it and replace the existing one.

**Comment by est31** [comment]
Thanks for your reply @spasche .

A few things. First, this is only a preview and a proof of concept. It needs lots of polishing and can't replace all functionalities of the old add on (like it still doesn't work with the file:// protocol). . Also, I know nothing about design in HTML and currently it looks really bad.

Second, I'm not really that keen on maintaining an add on. I just want to continue to use it. Of course, if doing the maintenance is too much for you I'll have to take over.

There are still some outstanding issues with web extension support in Firefox. Like right now it doesn't show a download in the download manager (probably a bug). There is also lacking support to put the web extension into the natural place, the actual "unknown mime type" dialog.

So until this add on stops working, I suggest to keep the pre web extension version and to use it instead. If the support has matured until then, I hope that the functionality can actually be replaced.

**Comment by neroux** [comment]
@spasche I can understand why you might not be motivated anymore, however it would be still nice if you could at least merge #26 as, regardless of Webextensions or not, OIB is already non-functional in current Firefox versions.

**Comment by spasche** [comment]
@neroux: sure, I'll have a look at it shortly (I commented on the PR).

File: background.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+browser.webRequest.onHeadersReceived.addListener(
+  headerRecv,
+  {urls: ["<all_urls>"]},
+  ["blocking", "responseHeaders"]
+);
+
+function handleMessage(data, sender, sendResponse) {
+	if (sender.url.startsWith(dispositionPage)) {
+		var tabId = sender.tab.id;
+		console.log("Message from the page: ",
```
Comment (Hassan3113) [inline background.js]: Text messages

Comment (Hassan3113) [inline background.js]: Find my messages

**Comment by Rick-74** [comment]
Any news on this?
The final deadline for non-WebExtensions addons is quickly approaching, and it would be a pity to lose such a useful one.

**Comment by est31** [comment]
@Rick-74 be assured that I'm intending to get a web extension version released (and hopefully also approved) before Firefox 57 gets released. If you are using nightly or beta I guess you'll have to disable the add on until then...

**Comment by HighCommander4** [comment]
> If you are using nightly or beta I guess you'll have to disable the add on until then...

Note that the original (non-WebExt) version should still work with Nightly if you flip the pref extensions.legacy.enabled to true.

**Comment by est31** [comment]
I've pushed an update with a few improvements. Now it works together with newer versions of Firefox.

Some areas for further improvement:

* Support displaying the dialog for invalid MIME types. This includes also the following point
* "Open as MIME type" feature
* Better UI design. The current one is crappy.
* Display the filename in the disposition page.
* If you click "Show native choice dialog" and then abort, you can't go back via the browser's history functionality.

Also, I've dug up [the place](https://dxr.mozilla.org/mozilla-central/source/uriloader/base/nsURILoader.cpp#604) where Firefox decides whether to invoke the dispatcher dialog or not, it might be of interest for development of the add-on as we basically need to replicate the behaviour (the linked `rv = helperAppService->DoContent(mContentType,` line invokes the dialog, the code above in the `DispatchContent` function handles the dispatching itself).

**Comment by HighCommander4** [comment]
Is this PR in a state where a prospective early adopter can try the modified plugin?

If so, could you mention how to build/install it? I looked at the README which says to run `make xpi` to build, but your commits remove the makefile.

**Comment by est31** [comment]
@HighCommander4 sure, try it out!

> could you mention how to build/install it?

You can install it for one restart, its explained in this video: https://www.youtube.com/watch?v=cer9EUKegG4

**Comment by HighCommander4** [comment]
I installed the extension as indicated, and tried opening https://bugs.eclipse.org/bugs/attachment.cgi?id=270041, which is a file with MIME type "text/x-c++src". However, I got the regular Open/Save dialog, with no "Open in browser" option.

**Comment by est31** [comment]
@HighCommander4 I've implemented a mode to override the sent MIME type, please test.

@spasche I feel its ready now. How should we manage the switch? Is there some way to transfer ownership on AMO? Or do you think its a better idea if I publish the web extension separately?

**Comment by est31** [comment]
All the points I've listed above are fixed now, except for the "browser's history" one. I'd add one point though, its recognizing when the browser can't handle the given MIME type, which is a bit tricky and I guess I'll need to default here and just let the user choose this via preferences.

**Comment by spasche** [comment]
Thanks for the work!

I can change the owner on AMO to hand it over to you, I just need the mail address from your AMO account. The AMO entry will also need some metadata updates.

We should make sure that the transition is smooth for users. I would suggest to first release a Beta version to get more testing: https://developer.mozilla.org/en-US/Add-ons/Distribution#Beta_versions

For the sources: if you're willing to maintain the add-on, it might make more sense that you host it on your own github account (i.e. https://github.com/est31/openinbrowser would be used as the project URL and for tickets).

**Comment by HighCommander4** [comment]
@est31 Thanks, the updated version works for me now!

Is it possible to package it as an XPI so it can be installed permanently?

**Comment by est31** [comment]
@spasche my account on AMO is https://addons.mozilla.org/de/firefox/user/est31/ and my e-mail address is included in [this link](https://github.com/est31/openinbrowser/commit/3cc7f7ac992bb77818e23097219e896eff12120b.patch).

Releasing a beta first sounds like a good idea.

**Comment by est31** [comment]
@HighCommander4 

> Is it possible to package it as an XPI so it can be installed permanently?

Creating an XPI is just [zipping up all files](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Publishing_your_WebExtension) and renaming the extension. However, Firefox stable requires the extension to be signed by Mozilla, which means upload and review to the Mozilla servers. You can disable this check for Firefox beta+nightly though (`xpinstall.signatures.required`). I guess the most convenient thing right now is to wait until I've uploaded the extension as beta on AMO and it passed review.

**Comment by spasche** [comment]
@est31: you should have access to the AMO project.

I gave it a try with Firefox 55 and 56.

If I open https://bugs.eclipse.org/bugs/attachment.cgi?id=270041, I see the "What to do with the file?" page. But if I click on the drop-down nothing happens and I see an error in the logs: Content Security Policy: The page’s settings blocked the loading of a resource at self (“script-src moz-extension://f1cdfa93-dccc-464f-8fe4-dd35d9d4ed15”).

If I open https://httpbin.org/response-headers?Server=httpbin&Content-Type=text%2Ffoo%3B+charset%3DUTF-8, the native "Opening ..." dialog appears.

**Comment by est31** [comment]
> But if I click on the drop-down nothing happens and I see an error in the logs

That error is not very important, it should be ignored. I've fixed it in either case.

I'm not sure why nothing happens when you click the drop down. At the very least the entry above should update. Does that happen? Do you have any other add-ons, maybe they do something? I confirm that one feature doesn't work on Firefox 55, its the "open page as source code" feature, not sure why.

> If I open [page with text/foo], the native "Opening ..." dialog appears.

Right, sadly there is no API to ask what Firefox would do with a given MIME type. I've implemented some basic heuristics but I guess I'll have to read firefox source code to improve them.

**Comment by est31** [comment]
> If I open [page with text/foo], the native "Opening ..." dialog appears.

Very interesting... it seems the page sends a duplicate Content-Type header. There is a [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1402570) in Firefox which makes it impossible to recognize the `text/foo` header here.

**Comment by est31** [comment]
Okay I've uploaded `2.0beta4` to AMO onto the beta channel. Please open any issues with the new add on on my github page https://github.com/est31/openinbrowser

**Comment by spasche** [comment]
@est31: my bad: I was expecting the add-on to show the content in the browser after selecting a mime in the list. If I click the "Open in Browser" button, it works. I don't know if it's just me or if the UI should be changed.

**Pull request closed without merging**